### PR TITLE
REST saga 2/3: Add REST endpoints to all RPCs

### LINKED ISSUE #4329: Moving rest-gateway http annotations to seperate files
Would it make sense to move the http annotations that are used in the proto definitions to seperate service configuration files as explained here:

https://grpc-ecosystem.github.io/grpc-gateway/docs/grpcapiconfiguration.html

https://cloud.google.com/endpoints/docs/grpc/grpc-service-config

I personally need to remove all http annotations when compiling to c#, as when compiling with the http annotations it adds a dependency to the google.api library, which I can't/don't want to include in my Unity projects.

### PR DESCRIPTION
**Pull request by guggero**
This is part 2 of 3 of the REST improvement saga which will ultimately replace #4141:

- [x] Part 1/3 (#4250): Cleanup proto files, split `WalletUnlocker` into own file, generate Swagger docs for all RPCs, fix comment styles (replace `///` with `//` and `/**` with `/*`).
- [x] Part 2/3 (This PR): Add REST endpoints and `grpc-gateway` code to all subservers and update `grpc-gateway` library.
- [ ] Part 3/3 (#4141): Add a WebSocket proxy to the REST listener to support streaming responses.

Fixes #4329.

### CODE REVIEW

**Comment by Roasbeef** [comment]
First part in, can be rebased now.

File: lnrpc/chainrpc/chainnotifier.proto [hunk 1]
```diff
@@ -14,7 +16,12 @@ service ChainNotifier {
     particular transaction by its hash or for an output script by specifying a
     zero hash.
     */
-    rpc RegisterConfirmationsNtfn (ConfRequest) returns (stream ConfEvent);
+    rpc RegisterConfirmationsNtfn (ConfRequest) returns (stream ConfEvent) {
+        option (google.api.http) = {
+            post: "/v2/chainnotifier/register_confirmations_ntfn"
```
Comment (Roasbeef) [inline lnrpc/chainrpc/chainnotifier.proto:21]: Could simplify these paths to: `register/confirmation`, `register/spend`, `register/blocks`.

File: lnrpc/walletrpc/walletkit.proto [hunk 2]
```diff
@@ -15,40 +16,68 @@ service WalletKit {
...
+        };
+    }
 
     /*
     NextAddr returns the next unused address within the wallet.
     */
-    rpc NextAddr (AddrRequest) returns (AddrResponse);
+    rpc NextAddr (AddrRequest) returns (AddrResponse) {
+        option (google.api.http) = {
+            get: "/v2/wallet/address/next"
```
Comment (Roasbeef) [inline lnrpc/walletrpc/walletkit.proto:42]: Since this modifies state, make it a POST?

File: lnrpc/walletrpc/walletkit.proto [hunk 3]
```diff
@@ -60,7 +89,11 @@ service WalletKit {
     remain supported. This is an advanced API that depends on the internals of
     the UtxoSweeper, so things may change.
     */
-    rpc PendingSweeps (PendingSweepsRequest) returns (PendingSweepsResponse);
+    rpc PendingSweeps (PendingSweepsRequest) returns (PendingSweepsResponse) {
+        option (google.api.http) = {
+            get: "/v2/wallet/sweeps"
```
Comment (Roasbeef) [inline lnrpc/walletrpc/walletkit.proto:94]: Latest version in master has a list _and_ a pending sweeps call.

File: lnrpc/watchtowerrpc/watchtower.proto [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
-    /* lncli: tower info
+    /* lncli: `tower info`
     GetInfo returns general information concerning the companion watchtower
     including its public key and URIs where the server is currently
     listening for clients.
     */
-    rpc GetInfo (GetInfoRequest) returns (GetInfoResponse);
+    rpc GetInfo (GetInfoRequest) returns (GetInfoResponse) {
+        option (google.api.http) = {
+            get: "/v2/watchtower"
```
Comment (Roasbeef) [inline lnrpc/watchtowerrpc/watchtower.proto:19]: Perhaps add a `/server` prefix here to disambiguate from the client calls below?

File: rpcserver.go [hunk 5]
```diff
@@ -766,18 +766,26 @@ func (r *rpcServer) Start() error {
...
 	err := lnrpc.RegisterLightningHandlerFromEndpoint(
-		context.Background(), mux, r.restProxyDest,
-		r.restDialOpts,
+		ctx, mux, r.restProxyDest, r.restDialOpts,
 	)
 	if err != nil {
 		return err
 	}
+
+	for _, subServer := range r.subServers {
```
Comment (Roasbeef) [inline rpcserver.go:779]: 👍

**Comment by Roasbeef** [comment]
Ready to leave draft now?

File: go.sum [hunk 6]
```diff
@@ -105,14 +107,18 @@ github.com/gogo/protobuf v1.1.1 h1:72R+M5VuhED/KujmZVcIquuo8mBgX4oVda//DQb3PXo=
...
 github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b h1:VKtxabqXZkF25pY9ekfRL6a582T4P37/31XEstQ5p58=
 github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b/go.mod h1:SBH7ygxi8pfUlaOkMMuAQtPIUF8ecWP5IEl/CR7VP2Q=
+github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e h1:1r7pUrabqp18hOBcwBwiTsbnFeTZHV9eER/QT5JVZxY=
 github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e/go.mod h1:cIg4eruTrX1D+g88fzRXU5OdNfaM+9IcxsU14FzY7Hc=
 github.com/golang/lint v0.0.0-20180702182130-06c8688daad7/go.mod h1:tluoj9z5200jBnyusfRPU2LqT6J+DAorxEvtC7LHB+E=
 github.com/golang/mock v1.1.1/go.mod h1:oTYuIxOrZwtPieC+H1uAHpcLFnEyAGVDL/k47Jfbm0A=
 github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
 github.com/golang/protobuf v1.3.1 h1:YF8+flBXS5eO826T4nzqPrxfhQThhXl0YzfuUPu4SBg=
 github.com/golang/protobuf v1.3.1/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
+github.com/golang/protobuf v1.3.2 h1:6nsPYzhq5kReh6QImI3k5qWzO4PEbvbIW2cwSfR/6xs=
```
Comment (Roasbeef) [inline go.sum:117]: `go mod tidy` to get rid of these duplicates?

File: lnrpc/routerrpc/router.proto [hunk 7]
```diff
@@ -40,49 +55,77 @@ service Router {
...
 
     /*
     QueryProbability returns the current success probability estimate for a
     given node pair and amount.
     */
     rpc QueryProbability (QueryProbabilityRequest)
-        returns (QueryProbabilityResponse);
+        returns (QueryProbabilityResponse) {
+        option (google.api.http) = {
+            get: "/v2/router/mc/probability/{from_node}/{to_node}/{amt_msat}"
```
Comment (Roasbeef) [inline lnrpc/routerrpc/router.proto:100]: 👍

File: go.sum (hunk 6 above)
Comment (guggero) [reply go.sum:116]: Not all of them. Even if I remove the lines with the old versions, a `go get ./...` adds them back in. I assume it's a dependency that itself depends on an older version of `protobuf`.

**Comment by guggero** [comment]
Reworked the PR to implement the suggestion in #4329 and added a very simple check to `make rpc-check` that makes sure every new RPC is added to the `lnrpc/rest-annotations.yaml` file.

File: Makefile [hunk 8]
```diff
@@ -210,6 +210,7 @@ rpc-format:
 
 rpc-check: rpc
 	@$(call print, "Verifying protos.")
+	for rpc in $$(find lnrpc/ -name "*.proto" | $(XARGS) awk '/    rpc /{print $$2}'); do if ! grep -q $$rpc lnrpc/rest-annotations.yaml; then echo "RPC $$rpc not added to lnrpc/rest-annotations.yaml"; exit 1; fi; done
```
Comment (cfromknecht) [inline Makefile:213]: 👍

**Pull request merged**